File: sample-chat-obj-c/sample-chat/Components/Authorization/AuthModule.m
//
//  AuthModule.m
//  sample-chat
//
//  Created by Injoit on 30.09.2022.
//  Copyright © 2022 Quickblox. All rights reserved.
//

#import "AuthModule.h"
#import <Quickblox/Quickblox.h>

NSString *const DEFAULT_PASSWORD = @"quickblox";

@implementation AuthModule
//MARK: - Public Methods
- (void)signUpWithFullName:(NSString *)fullName login:(NSString *)login {
    QBUUser *newUser = [[QBUUser alloc] init];
    newUser.login = login;
    newUser.fullName = fullName;
    newUser.password = DEFAULT_PASSWORD;
    
    __weak __typeof(self)weakSelf = self;
    [QBRequest signUp:newUser successBlock:^(QBResponse * _Nonnull response, QBUUser * _Nonnull user) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if ([strongSelf.delegate respondsToSelector:@selector(authModule:didSignUpUser:)]) {
            [strongSelf.delegate authModule:strongSelf didSignUpUser:user];
        }
        
    } errorBlock:^(QBResponse * _Nonnull response) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if (response.status == QBResponseStatusCodeValidationFailed) {
            // The user with existent login was created earlier
            if ([strongSelf.delegate respondsToSelector:@selector(authModule:didSignUpUser:)]) {
                [strongSelf.delegate authModule:strongSelf didSignUpUser:newUser];
            }
            return;
        }
        if ([strongSelf.delegate respondsToSelector:@selector(authModule:didReceivedError:)]) {
            [strongSelf.delegate authModule:strongSelf didReceivedError:response.error.error];
        }
    }];
}

- (void)loginWithFullName:(NSString *)fullName login:(NSString *)login {
    __weak __typeof(self)weakSelf = self;
    [QBRequest logInWithUserLogin:login
                         password:DEFAULT_PASSWORD
                     successBlock:^(QBResponse * _Nonnull response, QBUUser * _Nonnull user) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if ([strongSelf.delegate respondsToSelector:@selector(authModule:didLoginUser:)]) {
            [strongSelf.delegate authModule:strongSelf didLoginUser:user];
        }
    } errorBlock:^(QBResponse * _Nonnull response) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if ([strongSelf.delegate respondsToSelector:@selector(authModule:didReceivedError:)]) {
            [strongSelf.delegate authModule:strongSelf didReceivedError:response.error.error];
        }
    }];
}

- (void)updateFullName:(NSString *)fullName {
    QBUpdateUserParameters *updateUserParameter = [[QBUpdateUserParameters alloc] init];
    updateUserParameter.fullName = fullName;
    __weak __typeof(self)weakSelf = self;
    [QBRequest updateCurrentUser:updateUserParameter
                    successBlock:^(QBResponse * _Nonnull response, QBUUser * _Nonnull user) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if ([strongSelf.delegate respondsToSelector:@selector(authModule:didUpdateUpdateFullNameUser:)]) {
            [strongSelf.delegate authModule:strongSelf didUpdateUpdateFullNameUser:user];
        }
    } errorBlock:^(QBResponse * _Nonnull response) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if ([strongSelf.delegate respondsToSelector:@selector(authModule:didReceivedError:)]) {
            [strongSelf.delegate authModule:strongSelf didReceivedError:response.error.error];
        }
    }];
}

- (void)logout {
    __weak __typeof(self)weakSelf = self;
    [QBRequest logOutWithSuccessBlock:^(QBResponse * _Nonnull response) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if ([strongSelf.delegate respondsToSelector:@selector(authModuleDidLogout:)]) {
            [strongSelf.delegate authModuleDidLogout:strongSelf];
        }
    } errorBlock:^(QBResponse * _Nonnull response) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if ([strongSelf.delegate respondsToSelector:@selector(authModule:didReceivedError:)]) {
            [strongSelf.delegate authModule:strongSelf didReceivedError:response.error.error];
        }
    }];
}

@end

File: sample-chat-swift/sample-chat-swift/Components/Authorization/AuthModule.swift
//
//  AuthModule.swift
//  sample-chat-swift
//
//  Created by Injoit on 22.09.2022.
//  Copyright © 2022 quickBlox. All rights reserved.
//

import Foundation
import Quickblox

class ErrorInfo: NSObject {
    let info: String
    let statusCode: Int
    
    init(info: String, statusCode: Int) {
        self.info = info
        self.statusCode = statusCode
    }
}

@objc protocol AuthModuleDelegate: AnyObject {
    @objc optional func authModule(_ authModule: AuthModule, didSignUpUser user: QBUUser)
    @objc optional func authModule(_ authModule: AuthModule, didLoginUser user: QBUUser)
    @objc optional func authModule(_ authModule: AuthModule, didUpdateUpdateFullNameUser user: QBUUser)
    @objc optional func authModuleDidLogout(_ authModule: AuthModule)
    @objc optional func authModule(_ authModule: AuthModule, didReceivedError error: ErrorInfo)
}

class AuthModule: NSObject {
    //MARK: - Properties
    weak var delegate: AuthModuleDelegate?
    
    //MARK: - Public Methods
    func signUp(fullName: String, login: String) {
        let newUser = QBUUser()
        newUser.login = login
        newUser.fullName = fullName
        newUser.password = LoginConstant.defaultPassword
        QBRequest.signUp(newUser, successBlock: { [weak self] response, user in
            guard let self = self, let delegate = self.delegate else {
                return
            }
            delegate.authModule?(self, didSignUpUser: user)
            
        }, errorBlock: { [weak self] response in
            guard let self = self, let delegate = self.delegate else {
                return
            }
            if response.status == QBResponseStatusCode.validationFailed {
                // The user with existent login was created earlier
                delegate.authModule?(self, didSignUpUser: newUser)
                return
            }
            let error = ErrorInfo(info: response.localizedStatus, statusCode: response.status.rawValue)
            delegate.authModule?(self, didReceivedError: error)
        })
    }
    
    func login(fullName: String, login: String, password: String = LoginConstant.defaultPassword) {
        QBRequest.logIn(withUserLogin: login,
                        password: password,
                        successBlock: { [weak self] response, user in
            guard let self = self, let delegate = self.delegate else {
                return
            }
            delegate.authModule?(self, didLoginUser: user)

        }, errorBlock: { [weak self] response in
            guard let self = self, let delegate = self.delegate else {
                return
            }
            let error = ErrorInfo(info: response.localizedStatus, statusCode: response.status.rawValue)
            delegate.authModule?(self, didReceivedError: error)
        })
    }

    func updateFullName(fullName: String) {
        let updateUserParameter = QBUpdateUserParameters()
        updateUserParameter.fullName = fullName
        QBRequest.updateCurrentUser(updateUserParameter, successBlock: { [weak self] response, user in
            guard let self = self, let delegate = self.delegate else {
                return
            }
            delegate.authModule?(self, didUpdateUpdateFullNameUser: user)

        }, errorBlock: { [weak self] response in
            guard let self = self, let delegate = self.delegate else {
                return
            }
            let error = ErrorInfo(info: response.localizedStatus, statusCode: response.status.rawValue)
            delegate.authModule?(self, didReceivedError: error)
        })
    }

    func logout() {
        QBRequest.logOut(successBlock: { [weak self] response in
            guard let self = self, let delegate = self.delegate else {
                return
            }
            delegate.authModuleDidLogout?(self)
        }) {  [weak self] response in
            guard let self = self, let delegate = self.delegate else {
                return
            }
            let error = ErrorInfo(info: response.localizedStatus, statusCode: response.status.rawValue)
            delegate.authModule?(self, didReceivedError: error)
        }
    }
}
